direct.go

```diff
@@ -257,14 +308,34 @@ func (d *direct) RoundTrip(req *http.Request) (*http.Response, error) {
 			log.Debugf("Retrying domain-fronted request, pass %d", i)
 		}
 
-		req.Body = getBody()
-		conn, masqueradeGood, err := d.dial()
+		conn, m, masqueradeGood, err := d.dial()
 		if err != nil {
 			// unable to find good masquerade, fail
 			return nil, err
 		}
+		provider := d.providerFor(m)
+		if provider == nil {
+			log.Debugf("Skipping masquerade with disabled/unknown provider '%s'", m.ProviderID)
+			masqueradeGood(false)
+			continue
+		}
+		frontedHost := provider.Lookup(originHost)
+		if frontedHost == "" {
+			log.Tracef("Not translating unknown origin %s...", originHost)
+			frontedHost = originHost
+		} else {
+			log.Tracef("Translated origin %s -> %s for provider %s...", originHost, frontedHost, m.ProviderID)
+		}
+
+		reqi, err := cloneRequestWith(req, frontedHost, getBody())
+		if err != nil {
+			log.Errorf("Failed to copy http request?")
+			masqueradeGood(true)
+			continue
+		}
+
 		tr := httpTransport(conn, clientSessionCache)
-		resp, err := tr.RoundTrip(req)
+		resp, err := tr.RoundTrip(reqi)
 		if err != nil {
 			log.Debugf("Could not complete request %v", err)
 			masqueradeGood(false)
```

Comment:
On line 345, we check the status code against `http.StatusForbidden`. This is something we saw on CloudFront sometimes. 1) It's not clear if it applies to other providers and 2) It's possible other providers have their own idiosyncrasies. Perhaps we should make this configurable by provider somehow?

Reply:
That's fair, I like it.  I'll see what I can do.

Reply:
I think it would be acceptable to just have provider-specific verification functions that are baked into the code. I don't think we need fully declarative configuration for this. For forward-compatibility, we could have a default verifier if we see a provider ID that doesn't have a specific one.